socks/metrics.go

```diff
@@ -0,0 +1,69 @@
+package socks
+
+import (
+	"github.com/cybozu-go/usocksd/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	proxyElapsedHist = promauto.With(metrics.Registry).NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "process_seconds",
+		Help:      "provides the time elapsed, in seconds, between proxy start and end",
+	}, []string{"result"})
+	proxyRequestsInflightGauge = promauto.With(metrics.Registry).NewGauge(prometheus.GaugeOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "inflight_requests",
+		Help:      "provides the number of requests currently in-flight",
+	})
+	proxyBytesTxHist = promauto.With(metrics.Registry).NewHistogram(prometheus.HistogramOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "tx_bytes",
+		Help:      "bytes copied from the source connection to the destination connection",
+	})
+	proxyBytesRxHist = promauto.With(metrics.Registry).NewHistogram(prometheus.HistogramOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "rx_bytes",
+		Help:      "bytes copied from the destination connection to the source connection",
+	})
+
+	proxyCopyErrCount = promauto.With(metrics.Registry).NewCounterVec(prometheus.CounterOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "copy_errors_total",
+		Help:      "number of errors encountered copying from source/destination to destination/source",
+	}, []string{"from"})
+
+	proxyCopyElapsedHist = promauto.With(metrics.Registry).NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "proxy",
+		Name:      "copy_seconds",
+		Help:      "time spent copying from source/destination to destination/source",
+	}, []string{"from"})
+
+	authNegotiateCounter = promauto.With(metrics.Registry).NewCounterVec(prometheus.CounterOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "socks5",
+		Name:      "auth_negotiated_total",
+		Help:      "number of auth negotiation",
+	}, []string{"type", "reason", "result"})
+
+	addressReadCounter = promauto.With(metrics.Registry).NewCounterVec(prometheus.CounterOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "socks5",
+		Name:      "address_read_total",
+		Help:      "address read total count",
+	}, []string{"reason", "result"})
+
+	socksResponseCounter = promauto.With(metrics.Registry).NewCounterVec(prometheus.CounterOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "socks",
+		Name:      "responses_total",
```

Comment:
このメトリクスの値はリクエスト数と一致するはずなので、`requests_total` にしたほうが慣習的な気がします

Reply:
The number of requests and responses do not always seem to match.
This is because if the request is not a correct SOCKS request, these are dropped.
https://github.com/cybozu-go/usocksd/blob/03fb5bd4d4c5c5e61658581c1b3bee67e18be65e/socks/server.go#L128-L156

Do you think we should count the number of these incorrect requests?

Reply:
I recognized that the word "request" is ambiguous.
How about keeping the number of *connections* as a metric to clarify what counts?

Reply:
Does that mean that even invalid SOCKS protocol connections count?